RavenDB-21900 Ensure not releasing compressed document in case it is referenced.

maciejaszyk opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-21900 
https://issues.hibernatingrhinos.com/issue/RavenDB-21899

### Additional description
This PR fixes both issues (RavenDB-21900) and (RavenDB-21899) as they are related.
In the case of RavenDB-21900, the underlying issue was that compressed buffers were freed in document dispose after writing to a stream. Since the projection class has a cache for all loaded documents, we may have ended up with a reference to the document instance where `Data` (data from document) and decompression buffer (underlying in Voron part) were already released.
https://github.com/ravendb/ravendb/blob/a0b6b320ce3e774e950832e0bc4ed38fee9e61ce/src/Raven.Server/Documents/Queries/StreamDocumentQueryResult.cs#L19-L22

However, we need to keep track of all documents we have loaded from storage to release them, to avoid leaking memory (keeping in mind that stream queries are mostly used to get a lot of documents). This PR introduced a new concept of `QueriedDocument` which counts the references currently in use for each document to have an indicator when it is safe to release memory of each document.

### Type of change

- [x] Bug fix
- [ ] Regression bug fix
- [x] Optimization
- [ ] New feature

### How risky is the change?

- [ ] Low 
- [ ] Moderate 
- [x] High
- [ ] Not relevant

### Backward compatibility

- [x] Non breaking change
- [ ] Ensured. Please explain how has it been implemented?
- [ ] Breaking change
- [ ] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [x] No

### Documentation update

- [ ] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [x] No documentation update is needed 

### Testing by Contributor

- [x] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [x] It has been verified by manual testing

### Testing by RavenDB QA team

- [x] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [ ] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [x] No

### UI work

- [ ] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [x] No UI work is needed

maciejaszyk committed:
RavenDB-21900 When document is cached we cannot release decompression buffer.

maciejaszyk committed:
RavenDB-21899 Release memory in streaming query when loaded documents are removed from cache.

maciejaszyk committed:
RavenDB-21900 CloneWith should persist if storage can be released.

maciejaszyk committed:
RavenDB-21899
-Move projections to use generic document instead of `Document`.
-Track referenceCount for releasing the memory
-Prevent leak in case of  compressed db
- Release all related memory to the projected document

maciejaszyk commented:
test this in debug please

maciejaszyk committed:
RavenDB-21900 Return parent document in case when load gets empty string (indicates original document).

arekpalinski commented on src/Raven.Server/Documents/Document.cs:
```diff
@@ -13,6 +14,108 @@ public class TimeSeriesStream
         public string Key;
     }
+
+    public sealed class QueriedDocument : Document, IResetSupport<List<QueriedDocument>>
```
Let's move it to a separate file located somewhere in the area of query handling code.

maciejaszyk replied:
Sure, I left it here for development purposes only.

arekpalinski commented on src/Raven.Server/Documents/Document.cs:
```diff
@@ -13,6 +14,108 @@ public class TimeSeriesStream
+        
+        private QueriedDocument _parent;
+        private QueriedDocument _child;
+        private List<QueriedDocument> _childs;
```
`_referencedDocs` or `_references` maybe? Just to follow our terminology that we use. We usually say that if a document is loaded via a property of another document the it's "a referenced document"

arekpalinski commented on src/Raven.Server/Documents/Document.cs:
```diff
@@ -13,6 +14,108 @@ public class TimeSeriesStream
+        private int _refCount = 1;
+        
+        
+        private QueriedDocument _parent;
```
`_referencingDoc`, `_referencedVia` or `_loadedVia` maybe?

arekpalinski commented on src/Raven.Server/Documents/Document.cs:
```diff
@@ -13,6 +14,108 @@ public class TimeSeriesStream
+        private QueriedDocument _parent;
+        private QueriedDocument _child;
+        private List<QueriedDocument> _childs;
+        public bool CanDispose => _refCount <= 1;
```
Can it be 0 or negative? Is it valid then? Should we have debug assert for that?

maciejaszyk replied:
Added assertions.

arekpalinski commented on src/Raven.Server/Documents/Document.cs:
```diff
@@ -13,6 +14,108 @@ public class TimeSeriesStream
+            return cloned;
+        }
+
+        public void RemoveChild(QueriedDocument child)
```
See my above comments about the usage of "referenced" / "referencing" instead of "child" / "parent" terminology

arekpalinski commented on src/Raven.Server/Documents/Queries/Results/QueryResultRetrieverBase.cs:
```diff
@@ -69,11 +120,13 @@ public abstract class QueryResultRetrieverBase : IQueryResultRetriever
         private QueryTimingsScope _functionScope;
         private QueryTimingsScope _loadScope;
```
Please check the warning below - `The field 'QueryResultRetrieverBase<TDocument>._loadScope' is never used`

arekpalinski commented on src/Raven.Server/Documents/Queries/Results/QueryResultRetrieverBase.cs:
```diff
@@ -49,8 +100,8 @@ public abstract class QueryResultRetrieverBase : IQueryResultRetriever
 
-        private LruDictionary<string, Document> _loadedDocuments;
-        private Dictionary<string, Document> _loadedDocumentsByAliasName;
+        private LruDictionary<string, TDocument> _loadedDocumentMarshall;
```
marshall?

arekpalinski commented on src/Raven.Server/Documents/Queries/Results/QueryResultRetrieverBase.cs:
```diff
@@ -255,9 +358,24 @@ protected void FinishDocumentSetup(Document doc, Lucene.Net.Search.ScoreDoc scor
+                                    if (ReferenceEquals(nested, doc.Data) == false)
+                                    {
+                                        using (doc)
+                                            doc.CloneWith<TDocument>(_context, nested);
```
Isn't `doc = ...` missing here?

arekpalinski commented on src/Raven.Server/Documents/Queries/Results/QueryResultRetrieverBase.cs:
```diff
@@ -408,47 +530,34 @@ private bool TryExtractValueFromIndex(ref RetrieverInput retrieverInput, FieldsT
-                    }, null);
-
+                {
+                    using var _ = doc;
```
Sometimes you do `using var _ = doc;` sometimes 

```
using (doc)
{
}
```
Can we be consistent? (I prefer second syntax here because the scope is explicitly defined then)

arekpalinski replied:
I still see this inconsistency. `using (doc)` is simpler and more explicit IMO

arekpalinski commented on src/Raven.Server/Documents/Queries/Results/QueryResultRetrieverBase.cs:
```diff
@@ -807,9 +904,12 @@ protected bool TryGetValue(FieldsToFetch.FieldToFetch fieldToFetch, Document doc
                 {
 
                     TryGetValue(fieldToFetch.FunctionArgs[i], document, ref retrieverInput, indexFields, anyDynamicIndexFields, out _, out args[i], token);
+                    // if (args[i] is QueriedDocument d)
```
Please remove commented code

arekpalinski replied:
It's still here

arekpalinski replied:
And it's still here :)

arekpalinski commented on src/Raven.Server/Documents/Queries/Results/QueryResultRetrieverBase.cs:
```diff
@@ -1162,7 +1260,8 @@ private static bool TryGetTimeSeriesFunction(string methodName, Query query, out
         }
 
-        private bool TryGetFieldValueFromDocument(Document document, FieldsToFetch.FieldToFetch field, out object value)
+        private bool TryGetFieldValueFromDocument<TDocument>(TDocument document, FieldsToFetch.FieldToFetch field, out object value)
```
Please check the warning - `Type parameter 'TDocument' has the same name as the type parameter from outer type 'QueryResultRetrieverBase<TDocument>'`

arekpalinski commented on src/Raven.Server/Documents/Queries/Results/QueryResultRetrieverBase.cs:
```diff
@@ -1193,7 +1292,8 @@ private bool TryGetFieldValueFromDocument(Document document, FieldsToFetch.Field
         }
 
-        private static string GetIdFromDocument(Document document)
+        private static string GetIdFromDocument<TDocument>(TDocument document)
```
Warning

arekpalinski replied:
Still here

arekpalinski commented on src/Raven.Server/Utils/LruDictionary.cs:
```diff
@@ -30,6 +40,10 @@ public bool TryGetValue(TKey key, out TValue value)
         return true;
     }
 
+    public void MaybeClean()
```
Does it make sense to have it?

arekpalinski commented on src/Raven.Server/Utils/LruDictionary.cs:
```diff
@@ -52,13 +69,75 @@ public bool TryGetValue(TKey key, out TValue value)
+                                    previousValue.Dispose();
+                                }
+                                else
+                                    break;
```
Can we change it to 

```
if (previousValue.CanDispose == false)
   break;

_releaser?.ReleaseItem(valueAtNode.Value);
previousValue.Dispose();
```

arekpalinski commented on src/Raven.Server/Utils/LruDictionary.cs:
```diff
@@ -52,13 +69,75 @@ public bool TryGetValue(TKey key, out TValue value)
+        void ReleaseItem(TValue value);
+    }
+    
+    public class DocumentReleaser : ICacheReleaser<QueriedDocument>
```
Please talk to me directly about it

arekpalinski commented on src/Raven.Server/Utils/LruDictionary.cs:
```diff
@@ -52,13 +69,75 @@ public bool TryGetValue(TKey key, out TValue value)
                 if (_cache.Count >= _maxCapacity)
                 {
                     var removeKey = _list.Last!.Value;
+                    if (_cache.TryGetValue(removeKey, out var toRemove) && typeof(TValue) == typeof(QueriedDocument) && _releaser != null)
```
I don't like this check here: `typeof(TValue) == typeof(QueriedDocument)`.

maciejaszyk committed:
RavenDB-21900 Pass context from map-reduce retriever to avoid NRE, DirectGetInternal as base method to retrieve document from cache.

maciejaszyk committed:
RavenDB-21899 CollectionQuery will use QueriedDocument instead of Document.

maciejaszyk committed:
RavenDB-21899 Adjust code for edgecases.

maciejaszyk commented:
test this in debug please

maciejaszyk committed:
RavenDB-21899
- When PulseTransaction creates a new transaction, we need to clear the cache to avoid invalid pointers
- Move QueriedDocument to a dedicated file
- Optimise QueriedDocument list for single reference
- Rename variables within `QueriedDocument` to make them more meaningful
- Handling of non-existing documents in the cache for memory calls

maciejaszyk commented:
test this in debug please

maciejaszyk committed:
RavenDB-21899 Do not overwrite document in cache if we previously loaded the document but suspend the current document object.

arekpalinski commented on src/Raven.Server/Documents/Document.cs:
```diff
@@ -32,11 +34,10 @@ public class Document : IDisposable
         private bool _metadataEnsured;
         private bool _disposed;
-
+        public bool IsDisposed => _disposed;
```
I see it's used only in Debug.Assert - can we expose it only in DEBUG as `internal`? I don't want someone started to rely on it

arekpalinski replied:
Actually I'd prefer AssertNotDisposed() to be available in DEBUG and call it when necessary

maciejaszyk replied:
Good idea, done

arekpalinski commented on src/Raven.Server/Documents/Document.cs:
```diff
@@ -124,15 +125,25 @@ public Document Clone(JsonOperationContext context)
             {
                 Etag = Etag,
-                StorageId = StorageId,
+                StorageId = Voron.Global.Constants.Compression.NonReturnableStorageId, // this is basically in-memory copy, so we should care about storage...
```
should or should not care?

maciejaszyk replied:
No, the decompressed buffer is owned by the first instance of document pushed into the LRU cache. All copies are storage independent (because it's in-memory clone).

arekpalinski commented on src/Raven.Server/Documents/Indexes/Index.cs:
```diff
@@ -3300,9 +3300,12 @@ public IndexDefinition GetIndexDefinition()
                                             new QueryResultsIterationState(queryContext.Documents, DocumentDatabase.Configuration.Databases.PulseReadTransactionLimit));
+                                        enumerator = pulsedTransactionEnumerator;
+                                            
+                                         pulsedTransactionEnumerator.OnPulse += retriever.ClearCache;
```
Please add a comment explaining why we need it. It's a tricky case.

maciejaszyk replied:
Added

arekpalinski commented on src/Raven.Server/Documents/Queries/Results/GraphQueryResultRetriever.cs:
```diff
@@ -64,9 +62,12 @@ public override bool TryGetKeyCorax(IndexSearcher searcher, long id, out Unmanag
-        protected override Document LoadDocument(string id) => DocumentsStorage.Get(_context, id);
+        protected override Document LoadDocument(Document parent, string id, ref RetrieverInput retrieverInput)
+        {
+            return id == string.Empty 
```
Is it valid to pass empty string as ID to LoadDocument?

maciejaszyk replied:
Yes, it is:
https://github.com/ravendb/ravendb/pull/18132#discussion_r1531651071

arekpalinski commented on src/Raven.Server/Documents/Queries/Results/QueryResultRetrieverBase.cs:
```diff
@@ -116,25 +175,86 @@ protected void FinishDocumentSetup(Document doc, Lucene.Net.Search.ScoreDoc scor
+
+                if (typeof(TDocument) == typeof(Document) && doc is {IsDisposed: true})
+                {
+                    goto LoadDocument;
```
Do we really need to use `goto`?

maciejaszyk replied:
No, removed

arekpalinski commented on src/Raven.Server/Documents/Queries/Results/QueryResultRetrieverBase.cs:
```diff
@@ -255,9 +375,24 @@ protected void FinishDocumentSetup(Document doc, Lucene.Net.Search.ScoreDoc scor
+                                {
+                                    if (ReferenceEquals(nested, doc.Data) == false)
+                                    {
+                                        using (var _ = doc)
```
You can have `using (doc)`

maciejaszyk replied:
I've changed it into
```csharp
using (_ = doc)
   //operation
```

since pattern 
```csharp
using (doc)
    doc = operation
```

gets a warning and I wanted to avoid a lot of pragma disable.

arekpalinski commented on src/Raven.Server/Documents/Queries/Results/QueryResultRetrieverBase.cs:
```diff
@@ -490,34 +627,19 @@ private static void ThrowInvalidQueryBodyResponse(object fieldVal)
-            {
-                doc.Data = newData;
-            }
-
```
Please check below warnings:
- internal static FieldType GetFieldType(string field, Lucene.Net.Documents.Document indexDocument)
- internal class FieldType

arekpalinski commented on src/Raven.Server/Documents/Queries/Results/QueryResultRetrieverBase.cs:
```diff
@@ -1162,7 +1288,8 @@ private static bool TryGetTimeSeriesFunction(string methodName, Query query, out
         }
 
-        private bool TryGetFieldValueFromDocument(Document document, FieldsToFetch.FieldToFetch field, out object value)
+        private bool TryGetFieldValueFromDocument<TDocument>(TDocument document, FieldsToFetch.FieldToFetch field, out object value)
```
Please check the warning below

arekpalinski commented on src/Raven.Server/Documents/Queries/Results/QueryResultRetrieverBase.cs:
```diff
@@ -116,25 +175,86 @@ protected void FinishDocumentSetup(Document doc, Lucene.Net.Search.ScoreDoc scor
+        protected virtual TDocument LoadDocument(TDocument parentDocument, string id, ref RetrieverInput retrieverInput)
+        {
+            using var _ = (_loadScope = _loadScope?.Start() ?? _projectionScope?.For(nameof(QueryTimingsScope.Names.Load)));
+            if (id == string.Empty) // string.Empty means it's from original document.
```
It it a new concept or we had if before?

maciejaszyk replied:
Old one:
https://github.com/ravendb/ravendb/blob/2955874a713a26a8a57b787c17bdd1be0a4cb4dc/src/Raven.Server/Documents/Queries/Results/QueryResultRetrieverBase.cs#L888

arekpalinski commented on src/Raven.Server/Documents/Queries/QueriedDocument.cs:
```diff
@@ -0,0 +1,133 @@
+    public bool CanDispose => _refCount == 1;
+    public bool IsCloned;
+
+    public void LinkReferencedDocument(QueriedDocument additionalDocument)
```
`additionalDocument` is a document that is referenced right? Why not to call to `referencedDocument`?

maciejaszyk replied:
Done

arekpalinski commented on src/Raven.Server/Documents/Queries/QueriedDocument.cs:
```diff
@@ -0,0 +1,133 @@
+            _referencing ??= new();
+            _referencing.Add(_referencingSingle);
+            // ExplicitNullReference means we've multiple references.
+            _referencingSingle = ExplicitNullReference;
```
I understand the motivation of having `_referencingSingle` but I have doubts about the usage of `ExplicitNullReference;` as the marker and relaying on `ReferenceEquals()`

maciejaszyk replied:
Changed

arekpalinski commented on src/Raven.Server/Documents/Queries/QueriedDocument.cs:
```diff
@@ -0,0 +1,133 @@
+
+            Debug.Assert(ReferenceEquals(_sourceDocument.Data, Data) == false);
+            _sourceDocument.Dispose();
+            Debug.Assert(_sourceDocument.Data != null);
```
If any of those assertions fails one day during the tests you will not know which one. Please add messages as second parameter

maciejaszyk committed:
RavenDB-21899
 - Created QueriedDocumentCache to move logic with QueriedDocument from LRU Dictionary
 - Unified disposal logic
 - Added assert messages
 - Explained OnPulse functionality in comment

maciejaszyk commented:
test this in debug please

arekpalinski commented on src/Raven.Server/Documents/Queries/Results/QueriedDocumentCache.cs:
```diff
@@ -0,0 +1,85 @@
+                    {
+                        var element = toRemove.Value;
+                        if (element is {CanDispose: false})
+                            goto AddOnly;
```
Do we still have `goto` here?

maciejaszyk replied:
This is a different place, here it makes sense to have it.

arekpalinski commented on src/Raven.Server/Documents/Queries/Results/QueryResultRetrieverBase.cs:
```diff
@@ -116,25 +175,88 @@ protected void FinishDocumentSetup(Document doc, Lucene.Net.Search.ScoreDoc scor
+                    var parent = (QueriedDocument)(object)parentDocument;
+                    parent.LinkReferencedDocument(queriedDocument);
+                }
+            }
```
I wonder if we can do something with those checks:

```
if (typeof(TDocument) == typeof(QueriedDocument))
```

Let's talk directly about that.

arekpalinski commented on src/Raven.Server/Documents/Queries/Results/QueryResultRetrieverBase.cs:
```diff
@@ -255,9 +377,24 @@ protected void FinishDocumentSetup(Document doc, Lucene.Net.Search.ScoreDoc scor
+                                {
+                                    if (ReferenceEquals(nested, doc.Data) == false)
+                                    {
+                                        using (_ = doc)
```
You can drop `_ =`

arekpalinski commented on src/Raven.Server/Documents/Queries/Results/QueryResultRetrieverBase.cs:
```diff
@@ -255,9 +377,24 @@ protected void FinishDocumentSetup(Document doc, Lucene.Net.Search.ScoreDoc scor
                                 else if (fieldVal is Document d)
-                                    doc = d;
+                                {
+                                    using (_ = doc)
```
Same here

arekpalinski commented on src/Raven.Server/Documents/Queries/Results/QueryResultRetrieverBase.cs:
```diff
@@ -315,17 +456,40 @@ private bool TryExtractValueFromIndex(ref RetrieverInput retrieverInput, FieldsT
+                    Debug.Assert(documentFromCache.Data.BasePointer != null, "documentFromCache.Data.BasePointer != null");
+                }
+                
+                if (typeof(TDocument) == typeof(QueriedDocument))
```
Isn't better to do `if (documentFromCache is QueriedDocument queriedDocument)`?

arekpalinski commented on src/Raven.Server/Documents/Queries/Results/QueryResultRetrieverBase.cs:
```diff
@@ -490,34 +633,19 @@ private static void ThrowInvalidQueryBodyResponse(object fieldVal)
-                if (ReferenceEquals(newData, doc.Data) == false
-                    && doc.IgnoreDispose == false) // this is being referenced by the _loadedDocuments still...
-                    doc.Data?.Dispose();
+                using (_ = doc)
```
Please remove `_ =`

arekpalinski commented on src/Raven.Server/Documents/Queries/Results/QueryResultRetrieverBase.cs:
```diff
@@ -878,16 +964,25 @@ protected bool TryGetValue(FieldsToFetch.FieldToFetch fieldToFetch, Document doc
+                _loadedDocumentsByAliasName = new Dictionary<string, TDocument>();
             }
+
+            if (typeof(TDocument) == typeof(QueriedDocument) && _loadedDocumentIds != null)
```
Those checks are killing me: `typeof(TDocument) == typeof(QueriedDocument)`

arekpalinski commented on test/SlowTests/Issues/RavenDB-17134.cs:
```diff
@@ -13,15 +15,29 @@ public RavenDB_17134(ITestOutputHelper output) : base(output)
+            void AssertEqual(T other);
+        }
+        
+        private class TimeSeriesRange : IEqualsAssert<TimeSeriesRange>
```
Please check the warnings below

arekpalinski commented on test/SlowTests/Issues/RavenDB-17134.cs:
```diff
@@ -13,15 +15,29 @@ public RavenDB_17134(ITestOutputHelper output) : base(output)
         }
 
-        public class TimeSeriesRange
+        private interface IEqualsAssert<T>
```
Why do we need it?

maciejaszyk replied:
removed

arekpalinski commented on test/SlowTests/Issues/RavenDB_21900.cs:
```diff
@@ -0,0 +1,148 @@
+                   ModifyDatabaseRecord = x => x.DocumentsCompression = new DocumentsCompressionConfiguration(compressRevisions: true, compressAllCollections: true)
+               }))
+        {
+            string orderId;
```
Warning

arekpalinski commented on test/SlowTests/Issues/RavenDB_21900.cs:
```diff
@@ -0,0 +1,148 @@
+                   ModifyDatabaseRecord = x => x.DocumentsCompression = new DocumentsCompressionConfiguration(compressRevisions: true, compressAllCollections: true)
+               }))
+        {
+            string orderId;
```
Warning

maciejaszyk committed:
RavenDB-21899
- Asserts fix for map-reduce
-In case of fanout projection we need to increase reference counter to avoid disposing document.
-Avoid typeof(QueriedDocument) in QueryResultRetrieverBase
-Added implementation for increasing/linking documents to LruDictionaries to avoid type casts in QueryResultRetrieverBase.

maciejaszyk committed:
RavenDB-21900 Tests cleanup.

maciejaszyk commented:
test this in debug please

maciejaszyk committed:
RavenDB-21900 Add `TrackReferences` additional parameter.

ppekrol commented on test/SlowTests/Issues/RavenDB-17134.cs:
```diff
@@ -1,27 +1,44 @@
-
-        public class TimeSeriesRange
+        
+        private class TimeSeriesRange
```
warning here

ppekrol commented on test/SlowTests/Issues/RavenDB-17134.cs:
```diff
@@ -1,27 +1,44 @@
         }
 
-        class Dog
+        private class Dog
```
warning here

ppekrol commented on test/SlowTests/Issues/RavenDB-17134.cs:
```diff
@@ -31,10 +48,37 @@ public Dog(string name, int age)
+            }
         }
-        class User
+        private class User
```
warning here

maciejaszyk committed:
RavenDB-21900 Warning fix

maciejaszyk committed:
RavenDB-21900 When query has no load, use small cache to avoid persisting document too long.
